crates/astria-conductor/src/sequencer/mod.rs

```diff
@@ -186,36 +187,99 @@ impl Reader {
+                            let pool = pool.clone();
+                            resubscribe = subscribe_new_blocks(pool).boxed().fuse();
+                        }
                     }
                 }
             }
         }
     }
 }
 
-fn assign_new_blocks_subscription(
-    subscription: &mut stream::Fuse<NewBlocksStream>,
-    res: eyre::Result<NewBlocksStream>,
-) -> eyre::Result<()> {
-    use futures::stream::StreamExt as _;
-
-    let new_subscription = res.wrap_err("failed subscribing to new blocks from sequencer")?;
-    *subscription = new_subscription.fuse();
+#[derive(Debug, thiserror::Error)]
+enum ResubscriptionError {
+    #[error("failed getting a sequencer client from the pool")]
+    Pool(#[from] deadpool::managed::PoolError<crate::client_provider::Error>),
+    #[error("JSONRPC to subscribe to new blocks failed")]
+    JsonRpc(#[from] sequencer_client::extension_trait::SubscriptionFailed),
+    #[error("back off failed after 1024 attempts")]
+    BackoffFailed,
+}
 
-    Ok(())
+impl Clone for ResubscriptionError {
+    fn clone(&self) -> Self {
+        use deadpool::managed::{
+            HookError,
+            PoolError,
+        };
+        match self {
+            Self::Pool(e) => {
+                let e = match e {
+                    PoolError::Timeout(t) => PoolError::Timeout(*t),
+                    PoolError::Backend(e) => PoolError::Backend(e.clone()),
+                    PoolError::Closed => PoolError::Closed,
+                    PoolError::NoRuntimeSpecified => PoolError::NoRuntimeSpecified,
+                    PoolError::PostCreateHook(HookError::Message(s)) => {
+                        PoolError::PostCreateHook(HookError::Message(s.clone()))
+                    }
+                    PoolError::PostCreateHook(HookError::StaticMessage(m)) => {
+                        PoolError::PostCreateHook(HookError::StaticMessage(m))
+                    }
+                    PoolError::PostCreateHook(HookError::Backend(e)) => {
+                        PoolError::PostCreateHook(HookError::Backend(e.clone()))
+                    }
+                };
+                Self::Pool(e)
+            }
+            Self::JsonRpc(e) => ResubscriptionError::JsonRpc(e.clone()),
+            Self::BackoffFailed => ResubscriptionError::BackoffFailed,
+        }
+    }
 }
 
-async fn subscribe_new_blocks(pool: Pool<ClientProvider>) -> eyre::Result<NewBlocksStream> {
+async fn subscribe_new_blocks(
+    pool: Pool<ClientProvider>,
+) -> Result<NewBlocksStream, ResubscriptionError> {
     use sequencer_client::SequencerSubscriptionClientExt as _;
-    pool.get()
-        .await
-        .wrap_err("failed getting a sequencer client from the pool")?
-        .subscribe_new_block_data()
-        .await
-        .wrap_err("failed subscribing to sequencer to receive new blocks")
+    let retry_config = tryhard::RetryFutureConfig::new(1024)
```

Comment:
Should we use a more sophisticated less long pull retry stance here?

Like what about 10 retries on an exponential backoff starting at .5 and that maxes out at 10s? 10s is very slow when blocks are every 2s. You would defacto miss 5 blocks. 

```
tryhard::RetryFutureConfig::new(10)
    .exponential_backoff(std::time::Duration::from_millis(100))
    .max_delay(std::time::Duration::from_secs(10)
```

Reply:
Yeah, that makes sense. Pushed it this way. I don't know about the 10 retries. Effectively what it now does is 11 attempts (10 retries, starting from 100 ms exponential up to 10 s), and then it cycles again.